# Placement Prediction 🚀

## 1. Load Trained Model and Preprocessor

In this section, we load the fitted preprocessing pipeline and the selected machine learning model from the `models` directory.
These components will be used to transform new student data and generate placement predictions.

In [1]:
import pandas as pd
import joblib

In [2]:
preprocessor = joblib.load("../models/preprocessor.pkl")

In [3]:
best_model = joblib.load("../models/best_model.pkl")

In [4]:
type(preprocessor), type(best_model)

(sklearn.compose._column_transformer.ColumnTransformer,
 sklearn.ensemble._gb.GradientBoostingClassifier)

In [5]:
df = pd.read_csv("../data/processed/student_placement_features.csv")

In [6]:
df.columns.tolist()

['branch',
 'college_tier',
 'cgpa',
 'backlogs',
 'coding_skills',
 'dsa_score',
 'aptitude_score',
 'communication_skills',
 'ml_knowledge',
 'system_design',
 'internships',
 'projects_count',
 'certifications',
 'hackathons',
 'open_source_contributions',
 'extracurriculars',
 'placement_status',
 'technical_skill_score',
 'has_backlog',
 'experience_score',
 'technical_skill_gap']

## 2. Create New Student Input

A sample student profile is created using the original features collected from a student.

The engineered features will be calculated automatically using the same feature-engineering logic used during model development.

In [7]:
student_data = {
    "branch": "CSE",
    "college_tier": "Tier-1",
    "cgpa": 8.5,
    "backlogs": 0,
    "coding_skills": 8,
    "dsa_score": 8,
    "aptitude_score": 8,
    "communication_skills": 8,
    "ml_knowledge": 7,
    "system_design": 6,
    "internships": 2,
    "projects_count": 3,
    "certifications": 3,
    "hackathons": 2,
    "open_source_contributions": 2,
    "extracurriculars": 5
}

In [8]:
student_df = pd.DataFrame([student_data])
student_df

,branch,college_tier,cgpa,backlogs,coding_skills,dsa_score,aptitude_score,communication_skills,ml_knowledge,system_design,internships,projects_count,certifications,hackathons,open_source_contributions,extracurriculars
0,CSE,Tier-1,8.5,0,8,8,8,8,7,6,2,3,3,2,2,5


## 3. Apply Feature Engineering

The new student's original features are transformed using the same feature-engineering logic used during model development.

This ensures that the new input has the same feature structure as the data used to train the placement prediction model.

In [9]:
student_df["technical_skill_score"] = student_df[
    ["coding_skills", "dsa_score", "ml_knowledge", "system_design"]
].mean(axis=1)

In [10]:
student_df["experience_score"] = student_df[
    [
        "internships",
        "projects_count",
        "certifications",
        "hackathons",
        "open_source_contributions",
        "extracurriculars"
    ]
].sum(axis=1)

In [11]:
student_df["has_backlog"] = (
    student_df["backlogs"] > 0
).astype(int)

In [12]:
maximum_skill = student_df[
    ["coding_skills", "dsa_score", "ml_knowledge", "system_design"]
].max(axis=1)

minimum_skill = student_df[
    ["coding_skills", "dsa_score", "ml_knowledge", "system_design"]
].min(axis=1)

student_df["technical_skill_gap"] = maximum_skill - minimum_skill

In [13]:
student_df

,branch,college_tier,cgpa,backlogs,coding_skills,dsa_score,aptitude_score,communication_skills,ml_knowledge,system_design,internships,projects_count,certifications,hackathons,open_source_contributions,extracurriculars,technical_skill_score,experience_score,has_backlog,technical_skill_gap
0,CSE,Tier-1,8.5,0,8,8,8,8,7,6,2,3,3,2,2,5,7.25,17,0,2


## 4. Prepare Features for Prediction

The engineered student profile is prepared in the same feature structure used during model training.

The fitted preprocessing pipeline is then used to transform the categorical and numerical features into the format expected by the trained model.

In [14]:
student_encoded = preprocessor.transform(student_df)
student_encoded.shape

(1, 28)

## 5. Generate Placement Prediction

The processed student profile is passed to the trained Gradient Boosting model to predict the student's placement status.

The model returns:
- `1` → Placed
- `0` → Not Placed

In [15]:
prediction = best_model.predict(student_encoded)
prediction

array([1])

## 6. Convert Prediction to Placement Status

The numerical prediction returned by the model is converted into a human-readable placement status.

In [16]:
placement_status = "Placed" if prediction[0] == 1 else "Not Placed"
placement_status

'Placed'

## 7. Prediction Probability

The trained model's prediction probability is used to understand the model's confidence in the predicted placement status.

In [17]:
prediction_probability = best_model.predict_proba(student_encoded)
prediction_probability

array([[0.1137526, 0.8862474]])

## 8. Final Placement Prediction

The predicted placement status and its corresponding probability are combined into a user-friendly result.

In [18]:
placement_probability = prediction_probability[0][1] * 100
placement_probability

np.float64(88.62474035147827)

In [19]:
placement_probability = round(placement_probability, 2)
placement_probability

np.float64(88.62)

In [20]:
result = {
    "placement_status": placement_status,
    "placement_probability": placement_probability
}
result

{'placement_status': 'Placed', 'placement_probability': np.float64(88.62)}

## Placement Prediction

In this phase, we loaded the trained preprocessing pipeline and selected Gradient Boosting model.
A new student profile was created using the original student features.
The same feature-engineering steps used during model training were applied to the new student data.
The processed data was transformed using the fitted preprocessing pipeline and passed to the trained model.
The model predicted the student's placement status as **Placed**.
The model estimated an **88.62% placement probability** for the student.
Thus, the complete placement prediction pipeline was successfully implemented.


# 🎯 Career Recommendation

## Define Career Paths

Different career paths are defined based on the technical, analytical, and professional skills required for each role.

These career requirements will be used to compare the student's profile and determine the most suitable career path.

In [21]:
career_requirements = {
    "Data Scientist": {
        "coding_skills": 7,
        "dsa_score": 6,
        "ml_knowledge": 8,
        "projects_count": 3,
        "internships": 1,
        "aptitude_score": 7,
        "communication_skills": 6
    },

    "Data Analyst": {
        "coding_skills": 5,
        "dsa_score": 4,
        "ml_knowledge": 5,
        "projects_count": 2,
        "internships": 1,
        "aptitude_score": 8,
        "communication_skills": 8
    },

    "Machine Learning Engineer": {
        "coding_skills": 8,
        "dsa_score": 7,
        "ml_knowledge": 9,
        "projects_count": 3,
        "internships": 1,
        "aptitude_score": 6,
        "communication_skills": 6
    },

    "Software Developer": {
        "coding_skills": 8,
        "dsa_score": 8,
        "ml_knowledge": 5,
        "projects_count": 3,
        "internships": 1,
        "aptitude_score": 7,
        "communication_skills": 6
    },

    "Business Analyst": {
        "coding_skills": 4,
        "dsa_score": 3,
        "ml_knowledge": 4,
        "projects_count": 2,
        "internships": 1,
        "aptitude_score": 8,
        "communication_skills": 9
    }
}

career_requirements

{'Data Scientist': {'coding_skills': 7,
  'dsa_score': 6,
  'ml_knowledge': 8,
  'projects_count': 3,
  'internships': 1,
  'aptitude_score': 7,
  'communication_skills': 6},
 'Data Analyst': {'coding_skills': 5,
  'dsa_score': 4,
  'ml_knowledge': 5,
  'projects_count': 2,
  'internships': 1,
  'aptitude_score': 8,
  'communication_skills': 8},
 'Machine Learning Engineer': {'coding_skills': 8,
  'dsa_score': 7,
  'ml_knowledge': 9,
  'projects_count': 3,
  'internships': 1,
  'aptitude_score': 6,
  'communication_skills': 6},
 'Software Developer': {'coding_skills': 8,
  'dsa_score': 8,
  'ml_knowledge': 5,
  'projects_count': 3,
  'internships': 1,
  'aptitude_score': 7,
  'communication_skills': 6},
 'Business Analyst': {'coding_skills': 4,
  'dsa_score': 3,
  'ml_knowledge': 4,
  'projects_count': 2,
  'internships': 1,
  'aptitude_score': 8,
  'communication_skills': 9}}

## Create Student Profile

The student's profile used for placement prediction is reused for career recommendation.  
The profile contains academic performance, technical skills, professional experience, and other relevant attributes.

In [22]:
student_profile = {
    "coding_skills": 8,
    "dsa_score": 8,
    "ml_knowledge": 7,
    "projects_count": 3,
    "internships": 2,
    "aptitude_score": 8,
    "communication_skills": 8
}

student_profile

{'coding_skills': 8,
 'dsa_score': 8,
 'ml_knowledge': 7,
 'projects_count': 3,
 'internships': 2,
 'aptitude_score': 8,
 'communication_skills': 8}

## Calculate Career Suitability Scores

The student's profile is compared with the requirements of each career path.  
A suitability score is calculated based on how closely the student's skills and experience match the requirements of each career.  
The scores will be used to rank the available career paths.

In [23]:
career_scores = {}

for career, requirements in career_requirements.items():

    total_score = 0
    total_requirements = 0

    for feature, required_value in requirements.items():

        student_value = student_profile[feature]

        if feature in ["projects_count", "internships"]:
            student_value = min(student_value, required_value)

        score = min(student_value / required_value, 1) * 100

        total_score += score
        total_requirements += 1

    career_scores[career] = round(total_score / total_requirements, 2)

career_scores

{'Data Scientist': 98.21,
 'Data Analyst': 100.0,
 'Machine Learning Engineer': 96.83,
 'Software Developer': 100.0,
 'Business Analyst': 98.41}

## Weighted Career Scoring

Different skills have different levels of importance for different career paths.  
Therefore, weights are assigned to each requirement based on its relevance to the career.  
The weighted scores will provide a more realistic comparison and help identify the most suitable career path for the student.

In [24]:
career_weights = {
    "Data Scientist": {
        "coding_skills": 0.20,
        "dsa_score": 0.10,
        "ml_knowledge": 0.25,
        "projects_count": 0.20,
        "internships": 0.10,
        "aptitude_score": 0.10,
        "communication_skills": 0.05
    },

    "Data Analyst": {
        "coding_skills": 0.10,
        "dsa_score": 0.05,
        "ml_knowledge": 0.10,
        "projects_count": 0.15,
        "internships": 0.10,
        "aptitude_score": 0.25,
        "communication_skills": 0.25
    },

    "Machine Learning Engineer": {
        "coding_skills": 0.25,
        "dsa_score": 0.15,
        "ml_knowledge": 0.30,
        "projects_count": 0.15,
        "internships": 0.05,
        "aptitude_score": 0.05,
        "communication_skills": 0.05
    },

    "Software Developer": {
        "coding_skills": 0.30,
        "dsa_score": 0.25,
        "ml_knowledge": 0.05,
        "projects_count": 0.20,
        "internships": 0.05,
        "aptitude_score": 0.10,
        "communication_skills": 0.05
    },

    "Business Analyst": {
        "coding_skills": 0.05,
        "dsa_score": 0.05,
        "ml_knowledge": 0.05,
        "projects_count": 0.15,
        "internships": 0.10,
        "aptitude_score": 0.30,
        "communication_skills": 0.30
    }
}

career_weights

{'Data Scientist': {'coding_skills': 0.2,
  'dsa_score': 0.1,
  'ml_knowledge': 0.25,
  'projects_count': 0.2,
  'internships': 0.1,
  'aptitude_score': 0.1,
  'communication_skills': 0.05},
 'Data Analyst': {'coding_skills': 0.1,
  'dsa_score': 0.05,
  'ml_knowledge': 0.1,
  'projects_count': 0.15,
  'internships': 0.1,
  'aptitude_score': 0.25,
  'communication_skills': 0.25},
 'Machine Learning Engineer': {'coding_skills': 0.25,
  'dsa_score': 0.15,
  'ml_knowledge': 0.3,
  'projects_count': 0.15,
  'internships': 0.05,
  'aptitude_score': 0.05,
  'communication_skills': 0.05},
 'Software Developer': {'coding_skills': 0.3,
  'dsa_score': 0.25,
  'ml_knowledge': 0.05,
  'projects_count': 0.2,
  'internships': 0.05,
  'aptitude_score': 0.1,
  'communication_skills': 0.05},
 'Business Analyst': {'coding_skills': 0.05,
  'dsa_score': 0.05,
  'ml_knowledge': 0.05,
  'projects_count': 0.15,
  'internships': 0.1,
  'aptitude_score': 0.3,
  'communication_skills': 0.3}}

## Calculate Weighted Career Scores

The suitability score for each career is calculated using the defined career requirements and their corresponding weights.  
Higher weights are given to skills that are more important for a particular career.  
The final weighted score represents how well the student's profile matches each career path.

In [25]:
weighted_career_scores = {}

for career, requirements in career_requirements.items():

    weighted_score = 0

    for feature, required_value in requirements.items():

        student_value = student_profile[feature]

        if feature in ["projects_count", "internships"]:
            student_value = min(student_value, required_value)

        skill_score = min(student_value / required_value, 1) * 100

        weighted_score += skill_score * career_weights[career][feature]

    weighted_career_scores[career] = round(weighted_score, 2)

weighted_career_scores

{'Data Scientist': 96.88,
 'Data Analyst': 100.0,
 'Machine Learning Engineer': 93.33,
 'Software Developer': 100.0,
 'Business Analyst': 96.67}

## Rank Career Paths

The calculated weighted career scores are sorted in descending order to rank the career paths according to their suitability for the student.

In [26]:
ranked_careers = sorted(
    weighted_career_scores.items(),
    key=lambda x: x[1],
    reverse=True
)

ranked_careers

[('Data Analyst', 100.0),
 ('Software Developer', 100.0),
 ('Data Scientist', 96.88),
 ('Business Analyst', 96.67),
 ('Machine Learning Engineer', 93.33)]

## Final Career Recommendation

The ranked career paths are analyzed to identify the most suitable career for the student.  
If multiple careers have the same highest score, the system retains the tied careers instead of arbitrarily selecting one.  
The highest-ranked career is presented as the primary recommendation, while other careers with the same score are presented as alternative recommendations.

In [27]:
highest_score = ranked_careers[0][1]

top_careers = [
    career for career, score in ranked_careers
    if score == highest_score
]

primary_career = top_careers[0]

alternative_careers = top_careers[1:]

recommendation = {
    "primary_career": primary_career,
    "primary_score": highest_score,
    "alternative_careers": alternative_careers
}

recommendation

{'primary_career': 'Data Analyst',
 'primary_score': 100.0,
 'alternative_careers': ['Software Developer']}

## Career Explanation & Skill Gap Analysis

The recommended career is analyzed to identify the student's strengths and areas for improvement.  
Skills that meet the career requirements are considered strengths, while skills below the required level are identified as skill gaps.

In [28]:
recommended_requirements = career_requirements[primary_career]

career_strengths = []
skill_gaps = []

for feature, required_value in recommended_requirements.items():

    student_value = student_profile[feature]

    if student_value >= required_value:
        career_strengths.append(feature)
    else:
        skill_gaps.append({
            "skill": feature,
            "current_level": student_value,
            "required_level": required_value
        })

career_strengths, skill_gaps

(['coding_skills',
  'dsa_score',
  'ml_knowledge',
  'projects_count',
  'internships',
  'aptitude_score',
  'communication_skills'],
 [])

## Final Career Recommendation Result

The final career recommendation is generated by combining the ranked career scores, primary recommendation, alternative career options, strengths, and identified skill gaps.

In [29]:
career_result = {
    "recommended_career": primary_career,
    "career_suitability_score": highest_score,
    "alternative_careers": alternative_careers,
    "strengths": career_strengths,
    "skill_gaps": skill_gaps
}

career_result

{'recommended_career': 'Data Analyst',
 'career_suitability_score': 100.0,
 'alternative_careers': ['Software Developer'],
 'strengths': ['coding_skills',
  'dsa_score',
  'ml_knowledge',
  'projects_count',
  'internships',
  'aptitude_score',
  'communication_skills'],
 'skill_gaps': []}

## Career Recommendation 🎯

In this phase, career requirements and weights were defined for five different career paths.
The student's profile was compared with each career using weighted suitability scoring.
The career paths were ranked based on their suitability scores.
The system identified **Data Analyst** as the primary career recommendation with a **100% suitability score**.
**Software Developer** was identified as an alternative career with the same suitability score.
The student's strengths and skill gaps were also analyzed for the recommended career.
Thus, the career recommendation system was successfully implemented.
